File: src/main/java/com/untouchable/everytime/User/Controller/UserController.java
package com.untouchable.everytime.User.Controller;

import com.untouchable.everytime.User.DTO.PasswordChangeRequest;
import com.untouchable.everytime.User.DTO.UserDTO;
import com.untouchable.everytime.User.Service.UserService;
import io.swagger.v3.oas.annotations.Operation;
import io.swagger.v3.oas.annotations.responses.ApiResponse;
import io.swagger.v3.oas.annotations.tags.Tag;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.security.Principal;

@Tag(name = "2. User Service API", description = "유저 정보 관련 API")
@RestController
@RequiredArgsConstructor
@RequestMapping("/user")
public class UserController {

    private final UserService userService;

    @GetMapping
    @Operation(summary = "User Info 조회", description = "JWT로 인증 후, 해당하는 유저 정보 조회")
    @ApiResponse(responseCode = "200", description = "유저 정보 가져오기 성공")
    public ResponseEntity<UserDTO> info(
            Principal principal) {
        return ResponseEntity.ok(userService.getUserInfo(principal.getName()));
    }

    @PatchMapping("/password")
    @Operation(summary = "비밀번호 수정", description = "비밀번호 수정")
    public ResponseEntity<String> updatePassword(
            @RequestBody PasswordChangeRequest userChangePasswordDTO,
            Principal principal) {
        return ResponseEntity.ok(userService.updatePassword(principal.getName(), userChangePasswordDTO));
    }

    @PatchMapping("/email")
    @Operation(summary = "이메일 수정", description = "이메일 수정")
    public ResponseEntity<String> updateEmail(
            @RequestParam String email,
            Principal principal) {
        return ResponseEntity.ok(userService.updateEmail(principal.getName(), email));
    }

    @PatchMapping("/nickname")
    @Operation(summary = "닉네임 수정", description = "닉네임 수정")
    public ResponseEntity<String> updateNickname(
            @RequestParam String nickname,
            Principal principal) {
        return ResponseEntity.ok(userService.updateNickname(principal.getName(), nickname));
    }


    @DeleteMapping
    @Operation(summary = "회원 탈퇴", description = "회원 탈퇴 하는 기능")
    public ResponseEntity<String> delete(
            Principal principal) {
        return ResponseEntity.ok(userService.deleteUser(principal.getName()));
    }
}

File: src/main/java/com/untouchable/everytime/user/controller/UserController.kt
package com.untouchable.everytime.user.controller

import com.untouchable.everytime.common.domain.AuthenticatedUser
import com.untouchable.everytime.config.SwaggerConfig.BEARER_AUTH
import com.untouchable.everytime.user.Service.UserService
import com.untouchable.everytime.user.Service.data.UserData
import com.untouchable.everytime.user.controller.data.PasswordChangeForm
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.security.SecurityRequirement
import io.swagger.v3.oas.annotations.tags.Tag
import org.springframework.http.ResponseEntity
import org.springframework.security.core.annotation.AuthenticationPrincipal
import org.springframework.web.bind.annotation.DeleteMapping
import org.springframework.web.bind.annotation.GetMapping
import org.springframework.web.bind.annotation.PatchMapping
import org.springframework.web.bind.annotation.RequestBody
import org.springframework.web.bind.annotation.RequestMapping
import org.springframework.web.bind.annotation.RequestParam
import org.springframework.web.bind.annotation.RestController

@Tag(name = "2. User", description = "유저 정보 관련 API")
@RestController
@RequestMapping("/v1/api/user")
@SecurityRequirement(name = BEARER_AUTH)
class UserController(
    private val userService: UserService
) {

    @GetMapping
    @Operation(summary = "User Info 조회", description = "JWT로 인증 후, 해당하는 유저 정보 조회")
    fun getUserInfo(
        @AuthenticationPrincipal authenticatedUser: AuthenticatedUser
    ): ResponseEntity<UserData> {
        return ResponseEntity.ok(
            userService.findUserById(authenticatedUser.userId)
        )
    }

    @PatchMapping("/nickname")
    @Operation(summary = "User Nickname 변경", description = "JWT로 인증 후, 해당하는 유저의 닉네임 변경")
    fun changeNickname(
        authenticatedUser: AuthenticatedUser,
        @RequestParam nickname: String
    ): ResponseEntity<UserData> {
        return ResponseEntity.ok(
            userService.changeNickname(authenticatedUser.userId, nickname)
        )
    }

    @PatchMapping("/email")
    @Operation(summary = "User Email 변경", description = "JWT로 인증 후, 해당하는 유저의 이메일 변경")
    fun changeEmail(
        @AuthenticationPrincipal authenticatedUser: AuthenticatedUser,
        @RequestParam email: String
    ): ResponseEntity<UserData> {
        return ResponseEntity.ok(
            userService.changeEmail(authenticatedUser.userId, email)
        )
    }

    @PatchMapping("/password")
    @Operation(summary = "User Password 변경", description = "JWT로 인증 후, 해당하는 유저의 비밀번호 변경")
    fun changePassword(
        authenticatedUser: AuthenticatedUser,
        @RequestBody passwordChangeForm: PasswordChangeForm,
    ): ResponseEntity<Unit> {
        userService.changePassword(authenticatedUser.userId, passwordChangeForm)
        return ResponseEntity.ok().build()
    }

    @DeleteMapping
    @Operation(summary = "User 삭제", description = "JWT로 인증 후, 해당하는 유저 삭제")
    fun deleteUser(
        authenticatedUser: AuthenticatedUser
    ): ResponseEntity<Unit> {
        userService.deleteUser(authenticatedUser.userId)
        return ResponseEntity.ok().build()
    }
}
